_posts/2022-02-03-pixel-tests-update.md

```diff
@@ -0,0 +1,136 @@
+   realized it.
+
+ - There are some animations, and we need to make sure that we don't
+   take a pixel test mid-animation.
+
+ - When the Shell of Cockpit is changed, the content area for pages
+   like Networking might change size.  This is a problem for external
+   projects that don't control the version of the Cockpit shell.  We
+   worked around this by giving the content iframe a fixed size (by
+   making the browser window slightly bigger or smaller if necessary).
+
+ - We would miss failed pixels when a test ran into a known issue. A
+   known issue would cause the test to be skipped, and one would have
+   to look very closely to spot that it also had a failing pixel
+   test. This means that we sometimes forgot to update some reference
+   images, and would later be surprised by how we could have missed
+   them.
+
+ - As we change tests, we would accumulate unused reference images.
+   The bots now force us to clean up properly.
+
+ - Even though the actual DOM rendering is suprisingly deterministic,
+   one or two pixels in a million still come out slightly differently,
+   from time to time.  We have no better idea but to allow a low
+   number of imperfections per pixel test.
+
+With every change, we are more certain that we can get the pixel test
+costs under control.  We might never reach a point where we can stop
+tweaking the pixel test machinery, but the same is true about our CI
+testing machinery in general.
+
+## Why don't we hate them?
+
+I was expecting that adding pixel tests to Cockpit would be a hard
+sell, but we all seem to like them.  That's awesome!  What a team!
+
+I was afraid that pixel tests would be very unreliable because
+browsers would not be nearly pixel perfect enough.  But it turns out
+that they are very good at this, actually.  We had to switch off font
+hinting, and results are not consistent between running the tests
+locally and runnning them in the cloud, but pixels are almost never
+changing randomly from one run to the next.
+
+The unreliability comes mostly from not getting the DOM into the exact
+same state every time a given pixel test runs. This has always been a
+problem with our tests and we are pretty good at dealing with it.
+
+However, while the traditional tests were just code in the main repo,
+pixel tests are binary files in a Git submodule. Submodules can be
+very confusing, and we tried to hide most of it behind a script.  And
+Github is actually pretty good with images.
+
+One thing we have learned is to update reference images only very late
+in the life of a pull request. Resolving merge conflicts of binary
+files in a submodule is just not fun.
+
+I think we all like them because the value that we derive from pixel
+tests is well worth their cost. It feels awesome to officially replace
+an image of a sucky part of Cockpit with a nice one while fixing the
+code, and to show off a new feature with a couple of screenshots that
+you know will be kept from regressing.
```

Comment:
We can now also land PatternFly updates without a human visual inspection and rely on green tests instead, as by now they already do a better judgement than us humans could ever do.

Reply:
Added.